src/shred_collector/shred_tracker.zig

```diff
@@ -0,0 +1,240 @@
+const std = @import("std");
+const sig = @import("../lib.zig");
+
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+const Mutex = std.Thread.Mutex;
+
+const Slot = sig.core.Slot;
+
+const MAX_SHREDS_PER_SLOT: usize = sig.shred_collector.MAX_SHREDS_PER_SLOT;
+
+const MIN_SLOT_AGE_TO_REPORT_AS_MISSING: u64 = 200;
+
+pub const Range = struct {
+    start: usize,
+    end: ?usize,
+};
+
+pub const BasicShredTracker = struct {
+    logger: sig.trace.Logger,
+    mux: Mutex = .{},
+    /// The slot that this struct was initialized with at index 0
+    start_slot: ?Slot,
+    /// The oldest slot still being tracked, which hasn't yet been finished
+    current_bottom_slot: Slot,
+    /// The highest slot for which a shred has been received and processed successfully.
+    max_slot_processed: Slot = 0,
+    /// The highest slot that has been seen at all.
+    max_slot_seen: Slot = 0,
+    /// ring buffer
+    slots: [num_slots]MonitoredSlot = .{.{}} ** num_slots,
+
+    const num_slots: usize = 1024;
+
+    const Self = @This();
+
+    pub fn init(slot: ?Slot, logger: sig.trace.Logger) Self {
+        return .{
+            .start_slot = slot,
+            .current_bottom_slot = slot orelse 0,
+            .logger = logger,
+        };
+    }
+
+    pub fn maybeSetStart(self: *Self, start_slot: Slot) void {
+        if (self.start_slot == null) {
+            self.start_slot = start_slot;
+            self.current_bottom_slot = start_slot;
+        }
+    }
+
+    pub fn registerShred(
+        self: *Self,
+        slot: Slot,
+        shred_index: u64,
+    ) !void {
+        self.mux.lock();
+        defer self.mux.unlock();
+
+        self.maybeSetStart(slot);
+        self.max_slot_seen = @max(self.max_slot_seen, slot);
+        const monitored_slot = try self.getSlot(slot);
```

Comment:
I actually ran into forking pretty often in my testing. The code is written to handle it gracefully, but not intelligently. 

All the code does now is make sure that we've received all shreds from `0..last_shred`, where last_shred is the index of the final shred in the slot according to whichever fork has the least number of shreds in that slot. It doesn't care about the contents of the shred, just the index.

Let's say fork A says there are 13 shreds in slot 100, and fork B says there are 51 shreds in slot 100. We have received shreds 0..8 and 22..32 from fork B, and 4..13 from fork A. It doesn't get stuck waiting for shreds 33..51. It just assumes the shorter fork is sufficient. Also, it doesn't validate that it has received every shred on that fork's version of the slot. It just assumes it's fine to get at least one shred with a particular index.

So, when there are forks, it doesn't actually guarantee that it has received the full block that was selected by consensus for a given slot. But to accomplish this, you really need to track the different forks and identify which fork was selected as slots are rooted. As the blockstore is implemented, it will be able to handle this.